File: client-user/types/customer.ts
import { UserAddress } from "./user_address"

export type CustomerStatus = "active" | "new" | "blocked" | "inactive"

export interface Customer {
  _id: string
  fullName: string
  email: string
  phone: string
  orders: number
  spent: number
  joinDate: string | null
  lastLogin: string | null
  status: CustomerStatus
  avatar?: string
}

export interface CustomerSummary {
   "totalCustomers": number
   "activeCustomers": number
   "newCustomers": number
   "inactiveCustomers": number
}

export interface CustomerQuickView {
  _id: string
  fullName: string
  avatar?: string
  email: string
  phone: string
  orders: number
  spent: number
  status: CustomerStatus
  joinDate: string | null
  lastLogin: string | null
  addresses: UserAddress[]
}

// Địa chỉ khách hàng
export interface CustomerAddress {
  _id: string
  isDefault: boolean
  name: string
  phone: string
  province: string
  receiver: string
  street: string
  ward: string
}

// Dữ liệu chi tiết khách hàng
export interface CustomerDetail {
  _id: string;
  fullName: string;
  email: string;
  phone: string;
  avatar?: string | null;
  status: CustomerStatus;
  joinDate: string;       
  lastLogin?: string | null;  
  addresses: CustomerAddress[];
  totalOrders: number;
  totalSpent: number;
  avgOrderValue: number;
  lastOrderDate?: string | null; 
}

// Mỗi đơn hàng của khách
export interface CustomerOrder {
  orderCode: string;
  date: string; // ISO string
  total: number;
  status: "pending" | "confirmed" | "shipping" | "delivered" | "cancelled";
  items: number;
  paymentMethod: string;
}

export interface UpdateCustomerRequest {
  fullName?: string;
  email?: string;
  phone?: string;
  addresses?: {
    _id?: string;        // Nếu update address có sẵn
    name?: string;
    receiver?: string;
    phone?: string;
    street?: string;
    ward?: string;
    wardCode?: string;
    province?: string;
    provinceCode?: string;
    isDefault?: boolean;
  }[];
}

File: server/app/models/customer.py
from datetime import datetime
from app.utils.format_time import to_utc_iso

def map_status(user):
    if user.get("isBlocked"):
        return "blocked"
    if user.get("isVIP"):
        return "vip"
    if user.get("createdAt") and (datetime.utcnow() - user["createdAt"]).days <= 30:
        return "new"
    return "active"


def serialize_customer_admin(user):
    return {
        "id": str(user["_id"]),
        "fullName": user.get("fullName") or user.get("name") or "Khách hàng",
        "email": user.get("email"),
        "phone": user.get("phone"),
        "orders": user.get("totalOrders", 0),
        "spent": user.get("totalSpent", 0),
        "joinDate": to_utc_iso(user.get("createdAt")),
        "lastOrder": to_utc_iso(user.get("last_order_date")),
        "status": map_status(user),
        "avatar": user.get("avatar"),
    }